File: src/test/java/com/example/hexagonal/application/command/ConfirmAndCancelHandlerTest.java
package com.example.hexagonal.application.command;

import com.example.hexagonal.adapter.output.FixedIdGenerator;
import com.example.hexagonal.adapter.output.messaging.InMemoryEventPublisher;
import com.example.hexagonal.adapter.output.persistence.InMemoryOrderCommandRepository;
import com.example.hexagonal.adapter.output.persistence.InMemoryOrderQueryRepository;
import com.example.hexagonal.application.dto.CreateOrderInput;
import com.example.hexagonal.application.dto.OrderItemInput;
import com.example.hexagonal.domain.exception.OrderNotFoundException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.assertj.core.api.Assertions.*;

class ConfirmAndCancelHandlerTest {

    private InMemoryOrderCommandRepository commandRepo;
    private InMemoryOrderQueryRepository queryRepo;
    private InMemoryEventPublisher eventPublisher;
    private CreateOrderHandler createHandler;
    private ConfirmOrderHandler confirmHandler;
    private CancelOrderHandler cancelHandler;

    @BeforeEach
    void setUp() {
        commandRepo = new InMemoryOrderCommandRepository();
        queryRepo = new InMemoryOrderQueryRepository();
        eventPublisher = new InMemoryEventPublisher(commandRepo, queryRepo);
        createHandler = new CreateOrderHandler(commandRepo, eventPublisher, new FixedIdGenerator("order-1"));
        confirmHandler = new ConfirmOrderHandler(commandRepo, eventPublisher);
        cancelHandler = new CancelOrderHandler(commandRepo, eventPublisher);
    }

    private void createTestOrder() {
        createHandler.execute(new CreateOrderInput("c1", List.of(
                new OrderItemInput("p1", "鍵盤", 2500, 1)
        )));
    }

    @Nested
    @DisplayName("ConfirmOrderHandler")
    class ConfirmTests {

        @Test
        void 應該成功確認訂單() {
            createTestOrder();
            var output = confirmHandler.execute("order-1");
            assertThat(output.status()).isEqualTo("CONFIRMED");
            assertThat(output.totalAmount()).isEqualTo(2500);
        }

        @Test
        void 不存在的訂單應該拋出OrderNotFoundError() {
            assertThatThrownBy(() -> confirmHandler.execute("non-existent"))
                    .isInstanceOf(OrderNotFoundException.class)
                    .hasMessageContaining("找不到訂單");
        }

        @Test
        void 確認後讀取模型應該同步更新() {
            createTestOrder();
            confirmHandler.execute("order-1");
            var readModel = queryRepo.findById("order-1");
            assertThat(readModel).isPresent();
            assertThat(readModel.get().status()).isEqualTo("CONFIRMED");
        }
    }

    @Nested
    @DisplayName("CancelOrderHandler")
    class CancelTests {

        @Test
        void 應該成功取消訂單() {
            createTestOrder();
            var output = cancelHandler.execute("order-1", "客戶取消");
            assertThat(output.status()).isEqualTo("CANCELLED");
            assertThat(output.reason()).isEqualTo("客戶取消");
        }

        @Test
        void 不存在的訂單應該拋出OrderNotFoundError() {
            assertThatThrownBy(() -> cancelHandler.execute("non-existent", "reason"))
                    .isInstanceOf(OrderNotFoundException.class);
        }
    }
}

File: tests/application/ConfirmAndCancelHandler.test.ts
// ============================================================
// Command Handler 測試 - 確認與取消訂單
// ============================================================

import { CreateOrderHandler } from '../../src/application/commands/CreateOrderHandler';
import { ConfirmOrderHandler } from '../../src/application/commands/ConfirmOrderHandler';
import { CancelOrderHandler } from '../../src/application/commands/CancelOrderHandler';
import { InMemoryOrderCommandRepository } from '../../src/adapters/output/persistence/InMemoryOrderCommandRepository';
import { InMemoryOrderQueryRepository } from '../../src/adapters/output/persistence/InMemoryOrderQueryRepository';
import { InMemoryEventPublisher } from '../../src/adapters/output/messaging/InMemoryEventPublisher';
import { FixedIdGenerator } from '../../src/adapters/output/IdGeneratorAdapter';
import { OrderNotFoundError } from '../../src/domain/errors/DomainError';

describe('ConfirmOrderHandler', () => {
  let createHandler: CreateOrderHandler;
  let confirmHandler: ConfirmOrderHandler;
  let commandRepo: InMemoryOrderCommandRepository;
  let queryRepo: InMemoryOrderQueryRepository;
  let eventPublisher: InMemoryEventPublisher;

  beforeEach(() => {
    commandRepo = new InMemoryOrderCommandRepository();
    queryRepo = new InMemoryOrderQueryRepository();
    eventPublisher = new InMemoryEventPublisher(commandRepo, queryRepo);
    const idGenerator = new FixedIdGenerator('order-001');

    createHandler = new CreateOrderHandler(commandRepo, eventPublisher, idGenerator);
    confirmHandler = new ConfirmOrderHandler(commandRepo, eventPublisher);
  });

  it('應該成功確認訂單', async () => {
    await createHandler.execute({
      customerId: 'c1',
      items: [{ productId: 'p1', productName: '鍵盤', unitPrice: 2500, quantity: 1 }],
    });

    const result = await confirmHandler.execute('order-001');

    expect(result.orderId).toBe('order-001');
    expect(result.status).toBe('CONFIRMED');
    expect(result.totalAmount).toBe(2500);
  });

  it('不存在的訂單應該拋出 OrderNotFoundError', async () => {
    await expect(confirmHandler.execute('non-existent')).rejects.toThrow(OrderNotFoundError);
  });

  it('確認後讀取模型應該同步更新', async () => {
    await createHandler.execute({
      customerId: 'c1',
      items: [{ productId: 'p1', productName: '鍵盤', unitPrice: 2500, quantity: 1 }],
    });
    await confirmHandler.execute('order-001');

    const readModel = await queryRepo.findById('order-001');
    expect(readModel!.status).toBe('CONFIRMED');
  });
});

describe('CancelOrderHandler', () => {
  let createHandler: CreateOrderHandler;
  let cancelHandler: CancelOrderHandler;
  let commandRepo: InMemoryOrderCommandRepository;
  let queryRepo: InMemoryOrderQueryRepository;
  let eventPublisher: InMemoryEventPublisher;

  beforeEach(() => {
    commandRepo = new InMemoryOrderCommandRepository();
    queryRepo = new InMemoryOrderQueryRepository();
    eventPublisher = new InMemoryEventPublisher(commandRepo, queryRepo);
    const idGenerator = new FixedIdGenerator('order-001');

    createHandler = new CreateOrderHandler(commandRepo, eventPublisher, idGenerator);
    cancelHandler = new CancelOrderHandler(commandRepo, eventPublisher);
  });

  it('應該成功取消訂單', async () => {
    await createHandler.execute({
      customerId: 'c1',
      items: [{ productId: 'p1', productName: '鍵盤', unitPrice: 2500, quantity: 1 }],
    });

    const result = await cancelHandler.execute('order-001', '客戶要求取消');

    expect(result.orderId).toBe('order-001');
    expect(result.status).toBe('CANCELLED');
    expect(result.reason).toBe('客戶要求取消');
  });

  it('不存在的訂單應該拋出 OrderNotFoundError', async () => {
    await expect(cancelHandler.execute('non-existent', '理由')).rejects.toThrow(
      OrderNotFoundError,
    );
  });
});
